Issue 27 what spi

lawik opened the pull request:
Not sure if we have a good way of testing the chunking stuff since that's deep in the IO layer. Any thoughts?

Aside from that, would be good to have your eye on this before pushing it through. Probably makes sense to make 1.0.1 for this right?

lawik committed:
Experimental addition for Red wHAT display.

lawik committed:
Cleaned up configuration of yellow display or red what in rpihal.
Cleaned up and added additional tests for those cases.

lawik committed:
Preliminary chunking code.

BrynjarM committed:
Got it working on wHAT

lawik committed:
Merge pull request #31 from BrynjarM/issue-27-what-spi

Got it working on wHAT

lawik committed:
Minor cleanup to wHAT fixes.

BrynjarM reviewed (approved):
Tested and verified (ee68044fd3424cbd836034b8841229804e152f27) that it works  on my wHAT

nyaray commented on lib/hal/rpiio.ex:
```diff
@@ -34,6 +34,7 @@ defmodule Inky.RpiIO do
   @spi_speed_hz 488_000
   @spi_command 0
   @spi_data 1
+  @spi_chunk_size 4096
```
Should be suffixed with `bits` or `bytes` for clarity of intent.

nyaray commented on lib/hal/rpiio.ex:
```diff
@@ -93,12 +94,28 @@ defmodule Inky.RpiIO do
   defp spi_write(pins, data_or_command, values) when is_list(values) do
     :ok = GPIO.write(pins.dc_pid, data_or_command)
-    {:ok, <<_::binary>>} = SPI.transfer(pins.spi_pid, :erlang.list_to_binary(values))
+    spi_write(pins, data_or_command, :erlang.list_to_binary(values))
```
Does it not break tests that you added an unnecessary duplicated write to `dc`? Remove the GPIO write if this is going to rely on the later clause via the recursive call.

nyaray replied:
My bad, upon closer inspection, we didn't have a test for that.

nyaray commented on lib/hal/rpiio.ex:
```diff
@@ -93,12 +94,28 @@ defmodule Inky.RpiIO do
+          length = if left > @spi_chunk_size, do: @spi_chunk_size, else: left
+
+          {:ok, <<_::binary>>} = SPI.transfer(pins.spi_pid, :binary.part(value, offset, length))
+        end
```
* lines 108-117 should probably be in a separate function, named something like `spi_write_chunked`
* `x` should be named `chunk_index` and `parts`, `chunk_count`

nyaray commented on lib/hal/rpihal.ex:
```diff
@@ -158,14 +159,15 @@ defmodule Inky.RpiHAL do
-      state
-    end
-  end
+  # Set voltage of VSH and VSL on Yellow device
```
The comment should probably be copied to the red what function.

nyaray replied:
✓

nyaray committed:
27: address some review issues

- don't set dc pin high/low twice for SPI commands
- rename module attribute for SPI command chunking
- tidy up SPI chunking implementation a bit

nyaray committed:
27: address review issue

nyaray committed:
27: address credo issues

nyaray commented:
@BrynjarM would you mind verifying the latest commit on this branch?

Since you've got a wHAT, I'll also be asking you to check out a performance improvement I've been working at some point in the not too distant future, if that's ok!

BrynjarM commented:
@nyaray sorry for the delay, tested fe9bab3e268605ee39c54e097b31576915f94765 and it works on my wHAT

Sure, just ping me when the performance improvements are ready

JonRowe commented:
Any chance of releasing this 😬 (sorry to ask but I ran into the what bug too and then noticed it was fixed here 😂 )

lawik commented:
Ah, I'll take a look :)
